logstash-core/lib/logstash/pipeline.rb

```diff
@@ -94,60 +112,164 @@ def filters?
-      start_outputs
+      @outputs.each {|o| o.register }
+      @filters.each {|f| f.register}
+
+      pipeline_workers = safe_pipeline_worker_count
+      batch_size = @settings[:pipeline_batch_size]
+      batch_delay = @settings[:pipeline_batch_delay]
+      @logger.info("Starting pipeline",
+                   :id => self.pipeline_id,
+                   :pipeline_workers => pipeline_workers,
+                   :batch_size => batch_size,
+                   :batch_delay => batch_delay)
+
+      pipeline_workers.times do |t|
+        @worker_threads << Thread.new do
+          LogStash::Util.set_thread_name("[#{pipeline_id}]>worker#{t}")
+          worker_loop(batch_size, batch_delay)
+        end
+      end
     ensure
       # it is important to garantee @ready to be true after the startup sequence has been completed
       # to potentially unblock the shutdown method which may be waiting on @ready to proceed
       @ready.make_true
     end
+  end
 
-    @logger.info("Pipeline started")
-    @logger.terminal("Logstash startup completed")
+  # Main body of what a worker threda does
+  # Repeatedly takes batches off the queu, filters, then outputs them
+  def worker_loop(batch_size, batch_delay)
+    running = true
 
-    wait_inputs
+    while running
+      # To understand the purpose behind this synchronize please read the body of take_batch
+      input_batch, shutdown_received = @input_queue_pop_mutex.synchronize { take_batch(batch_size, batch_delay) }
+      running = false if shutdown_received
+      @events_consumed.increment(input_batch.size)
+
+      filtered = filter_batch(input_batch)
+      @events_filtered.increment(filtered.size)
+
+      output_batch(filtered)
 
-    if filters?
-      shutdown_filters
-      wait_filters
-      flush_filters_to_output!(:final => true)
+      inflight_batches_synchronize { set_current_thread_inflight_batch(nil) }
     end
+  end
 
-    shutdown_outputs
-    wait_outputs
+  def take_batch(batch_size, batch_delay)
+    batch = []
+    # Since this is externally synchronized in `worker_look` wec can guarantee that the visibility of an insight batch
+    # guaranteed to be a full batch not a partial batch
+    set_current_thread_inflight_batch(batch)
+
+    shutdown_received = false
+    batch_size.times do |t|
+      event = t==0 ? @input_queue.take : @input_queue.poll(batch_delay)
+      # Exit early so each thread only gets one copy of this
+      # This is necessary to ensure proper shutdown!
+      next if event.nil?
+
+      if event == LogStash::SHUTDOWN
+        shutdown_received = true
+        break
+      end
 
-    @logger.info("Pipeline shutdown complete.")
-    @logger.terminal("Logstash shutdown completed")
+      batch << event
+    end
 
-    # exit code
-    return 0
-  end # def run
+    [batch, shutdown_received]
+  end
 
-  def wait_inputs
-    @input_threads.each(&:join)
+  def filter_batch(batch)
+    batch.reduce([]) do |acc,e|
+      if e.is_a?(LogStash::Event)
```

Comment:
I may be forgetting, but why would `e` not be an Event? When? Should we comment/document when this occurs?